### ISSUE TITLE
Promotion does not happen when Jenkins idled

### ISSUE THREAD

**Issue by lordofthejars** [openshiftio/openshift.io#3356]
I created a new quickstart in openshift.io. The build started correctly, and I simply wait until Jenkins is idled. 

With Jenkins Idled, and the build stopped in approval step, I clicked on the Promote button and nothing happened. 

What I did after that was to go to jenkins.openshift.io and Jenkins was unidling at that time. I wait until it was unidled and inspecting the build log I could see that the "Promote" action was not sent to pipeline and Jenkisn was waiting for approval. So I push again the promote button, after Jenkins is unidled and then the build process continues normally and version is promoted.

Trying to gather more info but now it is failing for any other reasons not related with this.

**Comment by maxandersen** [comment]
@qodfathr this issue of promote button not working been observed in the past but never really identified - now we know its because promote button flow isn't waiting/triggering properly jenkins when it has become idled.
Since idling is  critical for scale i'm marking this as sev2/p0 as we need a better fix than just turning idling off.

**Comment by lordofthejars** [comment]
I am trying to reproduce again. I was able to check the logs from an unidled instance, so although the Jenkins instance success on the process of unidling, the process was not updated hence not promoted. 

But trying again now.

**Comment by qodfathr** [comment]
Thanks, @lordofthejars . Don't underestimate how long Jenkins needs to be running before it services the next build...it can be many minutes after the Jenkins pod starts.

**Comment by lordofthejars** [comment]
@qodfathr I tried again and it worked. What I have done is just run the build wait for one hour and then click the promote button. This night I will leave all night.

**Comment by kishansagathiya** [comment]
@lordofthejars @maxandersen I was trying to reproduce this issue.  
This is the request that fails every time I click on the promote button.
```
Request URL: https://forge.api.prod-preview.openshift.io/api/openshift/services/jenkins/ksagathi-preview-jenkins/job/kishansagathiya/job/myspace-preview/job/master/2/input/Proceed/proceedEmpty
Request Method: POST
Status Code: 500 Internal Server Error
```
However, the pipeline was still promoted despite the 500 error (i.e., Promotion worked for me)

**Comment by lordofthejars** [comment]
Trying to reproduce this problem I faced next problem, and it is that for any reason, the build didn't start. I went to OSO and I noticed that the Jenkins was idled, so the the build was never started. After waiting for 30 minutes I decided to click `Start 1 Pod` link in OSO. And what I got was:

```An error occurred scaling deployment config jenkins.
 Reason: deploymentconfigs "jenkins" is forbidden: User "asotobue" cannot update deploymentconfigs/scale in the namespace "asotobue-jenkins": User "asotobue" cannot update deploymentconfigs/scale in project "asotobue-jenkins"```

I am not sure if this helps or not but it is something I noticed.

**Comment by kishansagathiya** [comment]
@gastaldi Looking at https://github.com/openshiftio/openshift.io/issues/3356#issuecomment-385415317, is this an issue for launcher team ? 
cc/ @krishnapaparaju

**Comment by gastaldi** [comment]
@kishansagathiya it seems to me to be an issue for the Build team. Although forge.api.openshift.io is invoked, that code has zero relation to launcher, as described in https://github.com/openshiftio/openshift.io/issues/2093

**Comment by kishansagathiya** [comment]
@krishnapaparaju Nope. Who handles forge.api.openshift.io ?

**Comment by gastaldi** [comment]
@kishansagathiya see the error logs in https://errortracking.prod-preview.openshift.io/openshift_io/fabric8-launcher-backend/?query=is%3Aunresolved+jenkins

**Comment by bmicklea** [comment]
Can someone clarify whether this issues affects the OSIO user experience?  I see a reference to touching OSO directly (which isn't part of the happy path) and looking at a specific API response (also not part of the happy path).

So if this seems to work correctly in OSIO for the user I'd remove P0 label, if not then let's keep it.

**Comment by lordofthejars** [comment]
No, it is not reproducible. Yesterday worked meanwhile not on Friday.

Just to clarify that this promote button comes from OSIO site, but then I
switch to OSO to see what's happening

**Comment by bmicklea** [comment]
Thanks @lordofthejars. Any idea how we fix this?  I can see a non-trivial number of users pausing while using OSIO long enough to hit these kind of idle/unidle issues.

**Comment by lordofthejars** [comment]
@bmicklea Not on my side. No idea why this is happening.

**Comment by lordofthejars** [comment]
Ok I can confirm this way of reproduce, at least it happened to me twice:

1) With Jenkins unidled create a new project
2) The project will run until you get the manual step to promote to prod
3) Goto OSO and scale down Jenkins to 0
4) Push promote button

What you will observe is that after 5 minutes or so Jenkis is unidled and ready to serve requests but the promotion does not occurs. 

5) Push again Promote button (with Jenkins unidled) and then everything works as expected.

**Comment by maxandersen** [comment]
yes, so the api endpoint called is https://forge.api.openshift.io/api/openshift/services/jenkins/manderse-jenkins/job/maxandersen/job/reponame/job/master/1/input/Proceed/abort

My guess is this api is still talking directly to jenkins rather than through the proxy. 

@gastaldi @edewit any idea where this is handled now ?

**Comment by maxandersen** [comment]
found the code https://github.com/fabric8-launcher/launcher-backend/blob/master/addons/osio-addon/src/main/java/io/fabric8/launcher/osio/jenkins/JenkinsPipelineRegistry.java ...trying to trace it.

**Comment by gastaldi** [comment]
Yes, It doesn't seem to use the proxy:
https://github.com/fabric8-launcher/launcher-backend/blob/master/web/src/main/java/io/fabric8/launcher/web/api/OpenShiftResource.java#L164

Can someone provide a PR?

Em Sex, 4 de mai de 2018 18:25, Max Rydahl Andersen <
notifications@github.com> escreveu:

**Comment by maxandersen** [comment]
PR in progress to try fix it at https://github.com/fabric8-launcher/launcher-backend/pull/387

but its currently givin gerrors.

general idea is to have launcher use jenkins.openshift.io proxy intsead of direct; but for that to wok we also need some wait logic or some async handlingas it can take minutes before its ready.

**Comment by gastaldi** [comment]
PR reverted. 

Can someone who understands what needs to happen with Jenkins proxy please stand up?

**Comment by alexeykazakov** [comment]
@gastaldi https://github.com/fabric8-services/fabric8-jenkins-proxy/pull/254 is might be the reason why your PR didn't work.

**Comment by gastaldi** [comment]
@alexeykazakov worth a shot, do you want me to re-revert my PR once your PR is merged?

**Comment by alexeykazakov** [comment]
@gastaldi yes. We will need to re-revert your PR if/when this PR is merged.

**Comment by krishnapaparaju** [comment]
@bmicklea 'Promoting to production' at OSIO can fail if Jenkins Pod is either idled / in the process of getting unidled. @kishansagathiya  This require a new end point at Jenkins Proxy factoring in current state of Jenkins Pod with async events send to UI.

**Comment by lordofthejars** [comment]
I think there is one method which already returns the current state of
Jenkins

**Comment by maxandersen** [comment]
dont think we will be able to nor want to add more end points to jenkins proxy in rush imo.

can't we in short term additonal logic to frontend or the back end to wait  to increase chances of success ?

**Comment by maxandersen** [comment]
alternative is to idle jenkins much less agressively - like every 8 hrs instead of 45 min.

**Comment by pradeepto** [comment]
> alternative is to idle jenkins much less agressively - like every 8 hrs instead of 45 min.

@maxandersen Is that an acceptable alternative? That increases the costs, doesn't it?

**Comment by maxandersen** [comment]
> @maxandersen Is that an acceptable alternative? That increases the costs, doesn't it?

yes, but a shortime win if necessary. if we have cpu/ram free might as well use it.

**Comment by maxandersen** [comment]
thinking about it I actually think the absolute simplest fix/workaround is to add in the ui a step that checks jenkins.openshift.io (or direct jenkins) actually responds. If it does - just do as now; if not - ask user to retry a bit later (optimally have a retry loop that can be cancelled...)

**Comment by alexeykazakov** [comment]
@maxandersen can someone one just try to test https://github.com/fabric8-services/fabric8-jenkins-proxy/pull/254 and https://github.com/fabric8-launcher/launcher-backend/pull/387 properly first? :)

**Comment by alexeykazakov** [comment]
> can someone one just try to test fabric8-services/fabric8-jenkins-proxy#254 and fabric8-launcher/launcher-backend#387 properly first

OK. After a closer look I don't think it will solve the problem - https://github.com/fabric8-services/fabric8-jenkins-proxy/pull/254#issuecomment-386891201

**Comment by hrishin** [comment]
> thinking about it I actually think the absolute simplest fix/workaround is to add in the ui a step that checks jenkins.openshift.io (or direct jenkins) actually responds. If it does - just do as now; if not - ask user to retry a bit later (optimally have a retry loop that can be cancelled...

Need to add check here.

https://github.com/fabric8-ui/fabric8-ui/blob/800768435cbee7f029f79ef95d629346daf13ecc/src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts#L41

**Comment by kishansagathiya** [comment]
@maxandersen @alexeykazakov @krishnapaparaju 
How about adding logic in the UI to keep checking for whether the jenkins is up or not. And Whatever request that is being sent on clicking the promote button, send it once jenkins is up.
Checkout https://github.com/fabric8-services/fabric8-jenkins-proxy/blob/master/static/html/index.html#L33-L51
Clearly with this approach, promotion will take time equal to time taken to start jenkins

**Comment by maxandersen** [comment]
Yes, @kishansagathiya something like is what I was thinking about.

lets just not add some UI that *blocks* on this - but something you can cancel/close or retry.

**Comment by kishansagathiya** [comment]
> lets just not add some UI that blocks on this - but something you can cancel/close or retry.

Not sure what you mean here. Can you elaborate a bit?

**Comment by lordofthejars** [comment]
IMO what Max means is that user should be able to cancel the task of promotion if it takes so much and don't want to wait more time, or if it fails, show a message allowing user to cancel or retry.

**Comment by bmicklea** [comment]
@catrobson a UI change (although it may be minor) is being discussed here. Adding you so we get a quick check from UX.

**Comment by catrobson** [comment]
Let me see if I am following this.  

**Current behavior:**
User sees a step in the pipeline with promotion link, clicks the link.  
Expected result:  modal with accept/reject shows, they can click either and the system reacts accordingly
Actual result: Somethings things go out to lunch, we don't show a pop-up, or anything, other than a spinning wheel on the step?  Or is the spinning wheel in the modal?

**Proposed behavior:**
User sees a step in the pipeline with a promotion link, clicks the link.
IF Jenkins is unidled, show modal with accept/reject.  They can click either and the system reacts accordingly.
IF Jenkins is idled, show a modal with a loading state and some messaging about us starting up Jenkins, and their promotion request will be serviced in a minute.  Once it is up, the front-end will be checking for it to be ready and will ˆthenˆ push the command around wanting to set up promotion, and the user will see the accept/reject options.  They can click either, and the system reacts accordingly.

I'm not certain I have the timing correct for UI - whether this happens on load of the accept/reject modal like described above, or this is only an issue once the user presses accept/reject.  If the later we need a different proposed behavior.

**Comment by kishansagathiya** [comment]
@catrobson 

> ** Expected result: modal with accept/reject shows, they can click either and the system reacts accordingly
** Actual result: Somethings things go out to lunch, we don't show a pop-up, or anything, other than a spinning wheel on the step? Or is the spinning wheel in the modal?

On `View Pipelines` page result is Actual result is same as expected result when we click on `Promote`
However on `Analyze` page Actual result is different. When we click on `Promote`, it doesn't show a pop up or anything. If Jenkins is up it directly promotes. If not, It would sometimes wait for Jenkins to get up and promote and sometimes not promote at all.

The proposed behaviour seems like the right thing we should do, but the first priority should be to avoid the case where promotion is not happening at all. If we can (if time permits) we will definitely try to do the proposed behaviour. Keep in mind that this mostly needs to go in by today.
WDYT?

Also taggint @rohitkrai03

**Comment by kishansagathiya** [comment]
We had a failed attempt at this with https://github.com/fabric8-ui/fabric8-ui/pull/2889
The whole retry logic doesn't perform well. Proxy doesn't behave the way it is supposed to. Proxy was not written to serve this purpose, we reused current proxy flow to serve this purpose anyway and thus we ended up with a switch with many codes.
Anyway after a long discussion with @containscafeine @sbose78 @piyush1594 and others
We decided that it would be cleaner and better to just create a new endpoint at API router of proxy API router that would just tell if Jenkins is up or not given an access_token.

**Comment by sbose78** [comment]
We've hit a wall with the current approach.

Why?
Our current approach was to make an async call to Jenkins proxy, during "Promote" to check if jenkins is up - the call would also have unidled jenkins if needed. 

However, there is no REST API in Jenkins Proxy which returns a distinct HTTP response code to imply if jenkins is up. It always returns a 302 first. Hence, the user experience refactored  in fabric8-ui PR 2889 with this approach will not work

What next?
- We will need a REST API in jenkins proxy which returns a 200 if Jenkins is up, and a non-200 if Jenkins is not up but being idled.
- The above API needs to be used in the UI to determine if Jenkins is up at the time of hitting "Promote".
- We'll revert the UI commit https://github.com/fabric8-ui/fabric8-ui/pull/2889/files

Btw, Kishan and Shubham has already fixed the CORS issues with Jenkins proxy.



What about the workshop ?
This bug affects only those users who have started a build and have then taken a 45 minute break after the "Promote" stage was reached. So, workshop attendees aren't really affected unless they they take the 45 min break in the middle of an ongoing build.

**Comment by sthaha** [comment]
> We will need a REST API in jenkins proxy which returns a 200 if Jenkins is up, and a non-200 if Jenkins is not up ( and being unidled ).

Isn't it better to return the `state` of jenkins instead? e.g. api/jenkins/state/:namespace returns ` { data: { state: "running|starting|idled" } errors: [ {code: xxx, description: "foobar" } ]` as the api itself does succeed ?

/cc @kishansagathiya @sbose78

**Comment by kishansagathiya** [comment]
> state: "running|starting|idled"

This is definitely better

**Comment by gastaldi** [comment]
Can I ask you guys to also POST directly to Jenkins instead of using launcher-backend as a proxy? I have created [1]  for that matter, but CORS issues are preventing it to work as it should :cry: 

[1] https://github.com/fabric8-ui/fabric8-ui/pull/2895#issuecomment-387944649

**Comment by kishansagathiya** [comment]
@sthaha 
The issue is intermittent even when Jenkins is idle. Since the request to promote is not going through the proxy, it would be good to spend some time on how things are happening right now.

In the attempt that we made it failed, because jenkins.prod-preview.openshift.io did not get JSESSIONID cookie. I would rather investigate why this cookie is not being sent. 

Because even with the new endpoint even though we return correct status and make sure that jenkins is up, if the request to promote goes through jenkins-proxy we will end up facing the same cookie problem

**Comment by containscafeine** [comment]
Update:
- A new status API endpoint was introduced in Idler (merged) https://github.com/fabric8-services/fabric8-jenkins-idler/pull/238
- There is ongoing work in proxy to expose the above endpoint for UI to use
- There is an open fix for the CORS issue by @gastaldi at https://github.com/fabric8-services/fabric8-jenkins-proxy/issues/257, but we're investigating if this the right API to add CORS to in the proxy

**Comment by sthaha** [comment]
Update: 
> There is ongoing work in proxy to expose the above endpoint for UI to use

In order to support this in jenkins proxy, we will have to add another server to the  proxy to serve publicly accessible api (new route). The server should perform authentication and then communicate with the idler which is all going to take a while, so instead we think may be easier to add this to the api service (wit) which will act as a gateway for idler.

cc: @gastaldi @containscafeine @kishansagathiya @sbose78

**Comment by krishnapaparaju** [comment]
@invincibleJai please work with @kishansagathiya  to integrate UI side changes for routing UI requests (build specific) to Jenkins Proxy

**Comment by invincibleJai** [comment]
Task list from UI to accomodate above changes : 

- [ ] - introduce env in fabric8-ui to consume newly introduced endpoint to check jenkins status
- [ ] - Service in fabric8-ui 
- [ ] - Polling of api to check if jenkins is up and running
- [ ] - UI changes needed  to show a modal with a loading state and message "Starting up Jenkins, it may take few minutes"
- [ ] - UI changes to show modal Jenkins is unidled with accept/reject. They can click either and the system reacts accordingly
- [ ] - Add specs

**Comment by invincibleJai** [comment]
> Do you show this modal dialog when the user clicks on "Input required"?

yes

> I think we should improve the UX here. Wouldn't it be better to display the status in place rather than showing another modal dialog?

@sthaha sorry for the confusion, it won't be a different modal window. It would be the same modal window where currently we show promote/abort buttons, in case of Jenkins idled the same modal content would show loading state (spinner) and the message "Starting up Jenkins, it may take few minutes" along with greyed out/hidden Accept/Reject buttons. Once Jenkins start running it would show what is there currently i.e promote/abort.

If user dismiss it before Jenkins start running and reopens after a couple of minutes clicking `input required` it would show the modal with Promote/Abort or waiting for status depending on Jenkins State


Let me know if it's fine.

@catrobson your inputs would help here.

**Comment by sthaha** [comment]
> If user dismiss it before Jenkins start running and reopens after a couple of minutes clicking input required it would show the modal with Promote/Abort or waiting for status depending on Jenkins State

Here we are requiring user to poll i.e check if the jenkins has started, again and again and again... and that UX can be improved by showing the spinner right next to "input required" until it is up and running. That way once the workflow becomes 
1. user clicks on `[input requred]`
2. A dialog is shown " jenkins starting ... "
3. they dismiss the dialog and `[input required]` now becomes `[input required ↻]` 
4. when jenkins is running it becomes `[input required ✔]` or something better :)

**Comment by kishansagathiya** [comment]
@sthaha @invincibleJai 

Check https://github.com/openshiftio/openshift.io/issues/3356#issuecomment-387149597
We already know the proposed behavior from UX. I think what @invincibleJai is saying matches with what UX has suggested except that button remain grayed out(unclickable) until Jenkins is up. 
I like what UX has proposed compared to what @sthaha has proposed [here](https://github.com/openshiftio/openshift.io/issues/3356#issuecomment-396821795).

**Comment by invincibleJai** [comment]
Have handled the changes as suggested by UX as in https://github.com/openshiftio/openshift.io/issues/3356#issuecomment-396562994

This can be closed cc @kishansagathiya @piyush1594

**Comment by piyush1594** [comment]
Tested. Worked fine for me. Closing this @kishansagathiya @invincibleJai
